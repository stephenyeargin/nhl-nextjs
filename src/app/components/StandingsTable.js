import React from 'react';
import Link from 'next/link';
import { PropTypes } from 'prop-types';
import TeamLogo from '@/app/components/TeamLogo';
import { formatStat } from '@/app/utils/formatters';

import '@/app/assets/datatables.css';
import '@/app/components/StandingsTable.scss';

const StandingsTable = ({ standings }) => {
  const tableRows = standings
    .sort((a, b) => {
      if (a.wildcardSequence === b.wildcardSequence) {
        if (a.divisionAbbrev === b.divisionAbbrev) {
          return (a.divisionSequence > b.divisionSequence) ? 1 : -1;
        }

        // Sort alphabetically
        return (a.divisionAbbrev > b.divisionAbbrev) ? 1 : -1;
      }

      return (a.wildcardSequence > b.wildcardSequence) ? 1 : -1;
    });

  // If row[0].points is less than row[3].points, swap 0,1,2 and 3,4,5
  if (tableRows[0].points < tableRows[3].points) {
    const temp = tableRows[0];
    tableRows[0] = tableRows[3];
    tableRows[3] = temp;
    const temp2 = tableRows[1];
    tableRows[1] = tableRows[4];
    tableRows[4] = temp2;
    const temp3 = tableRows[2];
    tableRows[2] = tableRows[5];
    tableRows[5] = temp3;
  }

  const wildcardRankings = ['1', '2', '3', '1', '2', '3', 'WC1', 'WC2', '9', '10', '11', '12', '13', '14', '15', '16'];

  return (
    <div className="overflow-x-auto scrollbar-hidden">
      <table className="statsTable">
        <thead>
          <tr className="text-sm text-nowrap border bg-slate-200 dark:bg-slate-800">
            <th className="w-10 text-center"></th>
            <th className="text-center">Team</th>
            <th className="w-15 text-center">GP</th>
            <th className="w-15 text-center">W</th>
            <th className="w-15 text-center">L</th>
            <th className="w-15 text-center">OT</th>
            <th className="w-15 text-center">PTS</th>
            <th className="w-15 text-center">P%</th>
            <th className="w-15 text-center">RW</th>
            <th className="w-15 text-center">ROW</th>
            <th className="w-15 text-center">GF</th>
            <th className="w-15 text-center">GA</th>
            <th className="w-15 text-center">DIFF</th>
            <th className="w-15 text-center">HOME</th>
            <th className="w-15 text-center">AWAY</th>
            <th className="w-15 text-center">S/O</th>
            <th className="w-15 text-center">L10</th>
            <th className="w-15 text-center">STRK</th>
          </tr>
        </thead>
        <tbody>
          {tableRows.map((team, i) => (
            <tr key={team.teamAbbrev.default} className={[2, 5, 7].includes(i) ? 'border-double border-b-4' : ''}>
              <td className="text-center border ranking">{wildcardRankings[i]}</td>
              <td className="border p-1 text-center text-sm font-semibold">
                <Link href={`/team/${team.teamAbbrev.default}`} className={team.clinchIndicator === 'e' ? 'opacity-40' : ''}>
                  <div className="flex items-center gap-2">
                    <TeamLogo
                      src={team.teamLogo}
                      className="h-6 w-6 hidden sm:block"
                      alt="Logo"
                    />
                    <div className="block md:hidden">
                      {team.teamAbbrev.default}
                      {' '}
                      {team.clinchIndicator && (
                        <span>({team.clinchIndicator})</span>
                      )}
                    </div>
                    <div className="hidden md:block">
                      {team.teamName.default}
                      {' '}
                      {team.clinchIndicator && (
                        <span>({team.clinchIndicator})</span>
                      )}
                    </div>
                  </div>
                </Link>
              </td>
              <td className="text-center">{formatStat(team.gamesPlayed)}</td>
              <td className="text-center">{formatStat(team.wins)}</td>
              <td className="text-center">{formatStat(team.losses)}</td>
              <td className="text-center">{formatStat(team.otLosses)}</td>
              <td className="text-center">{formatStat(team.points)}</td>
              <td className="text-center">{formatStat(team.pointPctg,3)}</td>
              <td className="text-center">{formatStat(team.regulationWins)}</td>
              <td className="text-center">{formatStat(team.regulationPlusOtWins)}</td>
              <td className="text-center">{formatStat(team.goalFor)}</td>
              <td className="text-center">{formatStat(team.goalAgainst)}</td>
              <td className="text-center">{formatStat(team.goalDifferential)}</td>
              <td className="text-center">{formatStat(team.homeWins)}-{formatStat(team.homeLosses)}-{formatStat(team.homeOtLosses)}</td>
              <td className="text-center">{formatStat(team.roadWins)}-{formatStat(team.roadLosses)}-{formatStat(team.roadOtLosses)}</td>
              <td className="text-center">{formatStat(team.shootoutWins)}-{formatStat(team.shootoutLosses)}</td>
              <td className="text-center">{formatStat(team.l10Wins)}-{formatStat(team.l10Losses)}-{formatStat(team.l10OtLosses)}</td>
              <td className="text-center">{formatStat(team.streakCode)}{formatStat(team.streakCount)}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

StandingsTable.propTypes = {
  standings: PropTypes.arrayOf(PropTypes.object).isRequired,
};

export default StandingsTable;
