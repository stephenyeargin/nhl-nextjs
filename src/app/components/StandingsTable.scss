@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
    .standingsTable {
        @apply text-xs w-full table-auto border-collapse border;

        th {
            @apply p-2;
        }

        td {
            @apply p-2 border text-sm;
        }

        tr:nth-child(even) {
            @apply bg-slate-500/10;
        }
        td.sorting_1 {
            @apply bg-blue-100 text-black;
        }
    }
}