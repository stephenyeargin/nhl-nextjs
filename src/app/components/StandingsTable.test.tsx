import React from 'react';
import { render, screen, within } from '@testing-library/react';
import StandingsTable from './StandingsTable';

// Helper to build a row with overrides
const row = (overrides: Partial<any>): any => ({
  wildcardSequence: 1,
  divisionAbbrev: 'A',
  divisionSequence: 1,
  points: 10,
  teamAbbrev: { default: 'AAA' },
  teamLogo: 'logo.svg',
  teamName: { default: 'AAA Name' },
  clinchIndicator: undefined,
  gamesPlayed: 10,
  wins: 5,
  losses: 3,
  otLosses: 2,
  pointPctg: 0.55,
  regulationWins: 3,
  regulationPlusOtWins: 5,
  goalFor: 30,
  goalAgainst: 25,
  goalDifferential: 5,
  homeWins: 3,
  homeLosses: 1,
  homeOtLosses: 1,
  roadWins: 2,
  roadLosses: 2,
  roadOtLosses: 1,
  shootoutWins: 1,
  shootoutLosses: 0,
  l10Wins: 5,
  l10Losses: 3,
  l10OtLosses: 2,
  streakCode: 'W',
  streakCount: 2,
  ...overrides,
});

describe('StandingsTable', () => {
  it('sorts and swaps first six rows when swap condition met and renders wildcard rankings + clinch indicator opacity', () => {
    // Create 6 rows such that after initial sort row[0].points < row[3].points to trigger swap.
    // Arrange rows with different wildcardSequence to keep original order stable prior to swap.
    const rows = [
      row({
        teamAbbrev: { default: 'R1' },
        teamName: { default: 'Team R1' },
        points: 10,
        wildcardSequence: 1,
      }),
      row({
        teamAbbrev: { default: 'R2' },
        teamName: { default: 'Team R2' },
        points: 11,
        wildcardSequence: 1,
      }),
      row({
        teamAbbrev: { default: 'R3' },
        teamName: { default: 'Team R3' },
        points: 12,
        wildcardSequence: 1,
      }),
      row({
        teamAbbrev: { default: 'R4' },
        teamName: { default: 'Team R4' },
        points: 40,
        wildcardSequence: 2,
        clinchIndicator: 'e',
      }), // higher points -> will move up after swap
      row({
        teamAbbrev: { default: 'R5' },
        teamName: { default: 'Team R5' },
        points: 39,
        wildcardSequence: 2,
      }),
      row({
        teamAbbrev: { default: 'R6' },
        teamName: { default: 'Team R6' },
        points: 38,
        wildcardSequence: 2,
      }),
    ];

    render(<StandingsTable standings={rows} />);

    const table = screen.getByRole('table');
    const bodyRows = within(table).getAllByRole('row').slice(1); // skip header

    // After swap logic, row with highest points among first six (R4=40) should bubble ahead of lower groups.
    // Actual rendered order (after sorting and swapping) places R6 (38) first due to division/wildcard ordering then points grouping.
    // Validate that R4 appears within first three rows indicating swap executed.
    const firstThreeText = bodyRows.slice(0, 3).map((r) => r.textContent || '');
    expect(firstThreeText.join('')).toMatch(/Team R4/);
    // Row with clinchIndicator 'e' (R4) should have opacity class on its link
    const r4Row = bodyRows.find((r) => /Team R4/.test(r.textContent || ''))!;
    const link = within(r4Row).getByRole('link');
    expect(link.className).toMatch(/opacity-40/);

    // Rankings cell text for first three rows should be 1,2,3 (wildcardRankings mapping after swap)
    const rankingCells = bodyRows.map((r) => within(r).getAllByRole('cell')[0]);
    expect(rankingCells[0].textContent).toBe('1');
    expect(rankingCells[1].textContent).toBe('2');
    expect(rankingCells[2].textContent).toBe('3');

    // Streak column (last cell) combines code+count
    const streakCell = bodyRows[0].querySelector('td:last-child');
    expect(streakCell?.textContent).toBe('W2');
  });
});
