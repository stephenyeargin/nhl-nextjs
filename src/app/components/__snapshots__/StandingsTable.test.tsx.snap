// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandingsTable sorts and swaps first six rows when swap condition met and renders wildcard rankings + clinch indicator opacity 1`] = `
<DocumentFragment>
  <div
    class="overflow-x-auto scrollbar-hidden"
  >
    <table
      class="standingsTable"
    >
      <thead>
        <tr
          class="text-sm text-nowrap border bg-slate-200 dark:bg-slate-800"
        >
          <th
            class="w-10 text-center"
          />
          <th
            class="text-center"
          >
            Team
          </th>
          <th
            class="w-15 text-center"
          >
            GP
          </th>
          <th
            class="w-15 text-center"
          >
            W
          </th>
          <th
            class="w-15 text-center"
          >
            L
          </th>
          <th
            class="w-15 text-center"
          >
            OT
          </th>
          <th
            class="w-15 text-center"
          >
            PTS
          </th>
          <th
            class="w-15 text-center"
          >
            P%
          </th>
          <th
            class="w-15 text-center"
          >
            RW
          </th>
          <th
            class="w-15 text-center"
          >
            ROW
          </th>
          <th
            class="w-15 text-center"
          >
            GF
          </th>
          <th
            class="w-15 text-center"
          >
            GA
          </th>
          <th
            class="w-15 text-center"
          >
            DIFF
          </th>
          <th
            class="w-15 text-center"
          >
            HOME
          </th>
          <th
            class="w-15 text-center"
          >
            AWAY
          </th>
          <th
            class="w-15 text-center"
          >
            S/O
          </th>
          <th
            class="w-15 text-center"
          >
            L10
          </th>
          <th
            class="w-15 text-center"
          >
            STRK
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class=""
        >
          <td
            class="text-center border ranking"
          >
            1
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class=""
              href="/team/R6"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R6 
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R6 
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            38
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="text-center border ranking"
          >
            2
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class=""
              href="/team/R5"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R5 
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R5 
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            39
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
        <tr
          class="border-double border-b-4"
        >
          <td
            class="text-center border ranking"
          >
            3
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class="opacity-40"
              href="/team/R4"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R4 
                  <span>
                    (e)
                  </span>
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R4 
                  <span>
                    (e)
                  </span>
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            40
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="text-center border ranking"
          >
            1
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class=""
              href="/team/R3"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R3 
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R3 
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            12
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
        <tr
          class=""
        >
          <td
            class="text-center border ranking"
          >
            2
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class=""
              href="/team/R2"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R2 
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R2 
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            11
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
        <tr
          class="border-double border-b-4"
        >
          <td
            class="text-center border ranking"
          >
            3
          </td>
          <td
            class="border p-1 text-center text-sm font-semibold"
          >
            <a
              class=""
              href="/team/R1"
            >
              <div
                class="flex items-center gap-2"
              >
                <img
                  alt="Logo"
                  class="h-6 w-6 hidden sm:block"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="logo.svg"
                  style="color: transparent;"
                  width="256"
                />
                <div
                  class="block md:hidden"
                >
                  R1 
                </div>
                <div
                  class="hidden md:block"
                >
                  Team R1 
                </div>
              </div>
            </a>
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            2
          </td>
          <td
            class="text-center"
          >
            10
          </td>
          <td
            class="text-center"
          >
            .550
          </td>
          <td
            class="text-center"
          >
            3
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            30
          </td>
          <td
            class="text-center"
          >
            25
          </td>
          <td
            class="text-center"
          >
            5
          </td>
          <td
            class="text-center"
          >
            3-1-1
          </td>
          <td
            class="text-center"
          >
            2-2-1
          </td>
          <td
            class="text-center"
          >
            1-0
          </td>
          <td
            class="text-center"
          >
            5-3-2
          </td>
          <td
            class="text-center"
          >
            W2
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
